src/include/migraphx/generic_float.hpp

```diff
@@ -0,0 +1,392 @@
+
+    static constexpr int exponent_bias()
+    {
+        return all_ones<ExponentSize - 1>();
+    }
+
+    explicit constexpr generic_float(float f = 0.0) noexcept
+    {
+        from_float(get_parts(f));
+    }
+
+    constexpr generic_float &operator=(float f) noexcept
+    {
+        from_float(get_parts(f));
+        return *this;
+    }
+
+    constexpr generic_float operator-() const noexcept
+    {
+        generic_float result = *this;
+        result.sign = !this->sign;
+        return result;
+    }
+
+    constexpr float to_float() const noexcept
+    {
+        float32_parts f{};
+
+        f.sign = sign;
+
+        if(exponent == 0)
+        {
+
+            if(mantissa == 0)
+            {
+
+                f.exponent = 0;
+                f.mantissa = 0;
+            }
+            else
+            {
+                int shift  = 0;
+                f.mantissa = mantissa;
+
+                while((f.mantissa & (1 << MantissaSize)) == 0)
+                {
+                    f.mantissa <<= 1;
+                    shift++;
+                }
+
+                f.mantissa &= all_ones<MantissaSize>();
+
+                f.exponent = float32_parts::exponent_bias() - exponent_bias() - shift + 1;
+                f.mantissa = f.mantissa << (float32_parts::mantissa_width() - MantissaSize);
+            }
+        }
+        else if(exponent == all_ones<ExponentSize>())
+        {
+            f.mantissa = mantissa << (float32_parts::mantissa_width() - MantissaSize);
+            f.exponent = float32_parts::max_exponent();
+        }
+        else
+        {
+            f.mantissa               = mantissa << (float32_parts::mantissa_width() - MantissaSize);
+            constexpr const int diff = float32_parts::exponent_bias() - exponent_bias();
+            f.exponent               = int(exponent) + diff;
+        }
+        return f.to_float();
+    }
+
+    constexpr void from_float(float32_parts f) noexcept
+    {
+        sign = f.sign;
+
+        if(f.exponent == 0)
+        {
+            exponent = 0;
+            mantissa = f.mantissa >> (float32_parts::mantissa_width() - MantissaSize);
+        }
+        else if(f.exponent == float32_parts::max_exponent())
+        {
+            exponent = all_ones<ExponentSize>();
+            mantissa = f.mantissa >> (float32_parts::mantissa_width() - MantissaSize);
+        }
+        else
+        {
+            constexpr const int diff = float32_parts::exponent_bias() - exponent_bias();
+            auto e = int(f.exponent) - diff;
+
+            if(e >= static_cast<int>(all_ones<ExponentSize>()))
+            {
+                exponent = all_ones<ExponentSize>();
+                mantissa = 0;
+            }
+            else if(e <= 0)
+            {
+                exponent = 0;
+
+                auto shift = diff - int(f.exponent);
+                mantissa =
+                    (f.mantissa | (1 << static_cast<int>(float32_parts::mantissa_width()))) >>
```

Comment:
This is the correct way to convert a subnormal fp. It is a 2-step process:

- Add the implicit leading 1 `(f.mantissa | (1 << static_cast<int>(float32_parts::mantissa_width())))` required for subnormal representations
- Shifting the mantissa left by shift to account for the smaller FP16 exponent range (`shift`), truncating the mantissa (`float32_parts::mantissa_width() - MantissaSize`). Additional shift by `1` bit to account for the added leading 1 in the subnormal case. Total shift: ` shift + float32_parts::mantissa_width() - MantissaSize + 1`

Subnormal numbers have no implicit 1 and represent values of the form 0.m rather than 1.m.